subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/ImplementationChanges.java

```diff
@@ -50,42 +49,34 @@ public boolean accept(ChangeVisitor visitor) {
                 executable.getDisplayName(), previousImplementation.getTypeName(), currentImplementation.getTypeName()));
         }
         if (currentImplementation.isUnknown()) {
-            return visitor.visitChange(new DescriptiveChange("The type of %s %s",
-                    executable.getDisplayName(), currentImplementation.getUnknownReason()));
+            // We already validate that the current implementation can't be unknown.
+            throw new GradleException("Cannot determine changes for work with unknown implementation");
```

Comment:
Is there a more specific exception to throw here?

Where does the message end up? I don't see it in tests.

Reply:
I removed the throwing of the exception and a longer comment about what is going on at the top of the method.